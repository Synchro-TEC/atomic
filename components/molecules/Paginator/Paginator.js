import React from 'react';
import PropTypes from 'prop-types';
import nanoid from 'nanoid';
import LinkButton from '../../atoms/LinkButton/LinkButton';

const Paginator = ({ page, limit, total, onGoToNext, onGoToPrevious, onGoToPage }) => {
  let currentPage = page;
  let totalOfPieces = Math.ceil(total / limit);

  const mountPaginateInfo = () => {
    return {
      limit: limit,
      offset: limit * currentPage - limit,
      currentPage: currentPage,
    };
  };

  const goToNext = () => {
    if (onGoToNext && currentPage < totalOfPieces) {
      currentPage += 1;
      onGoToNext(mountPaginateInfo());
    }
  };

  const goToPrevious = () => {
    if (currentPage > 1) {
      currentPage -= 1;
    }
    onGoToPrevious(mountPaginateInfo());
  };

  const goToPage = e => {
    currentPage = parseInt(e.target.textContent);
    onGoToPage(mountPaginateInfo());
  };

  const colorPaginationOptionAndPutGapIfNeeded = i => {
    let option = {};

    if (i === '...') {
      option = <span key={nanoid()}>{i}</span>;
    } else {
      if (currentPage === i) {
        option = <em key={nanoid()}> {i} </em>;
      } else {
        option = (
          <a key={nanoid()} onClick={e => goToPage(e)}>
            {' '}
            {i}{' '}
          </a>
        );
      }
    }

    return option;
  };

  const getGap = () => {
    return colorPaginationOptionAndPutGapIfNeeded('...');
  };

  const createInitialOptionPages = () => {
    let initialPages = [];
    if (totalOfPieces >= 2) {
      for (let i = 1; i <= 2; i++) {
        initialPages.push(colorPaginationOptionAndPutGapIfNeeded(i));
      }
    } else if (totalOfPieces > 0) {
      initialPages.push(colorPaginationOptionAndPutGapIfNeeded(totalOfPieces));
    }

    return initialPages;
  };

  const createMiddleOptionPages = () => {
    let middlePages = [];

    if (totalOfPieces <= 10) {
      for (let i = totalOfPieces - 2; i >= 3; i--) {
        middlePages.unshift(colorPaginationOptionAndPutGapIfNeeded(i));
      }
    } else if (totalOfPieces - currentPage <= 2) {
      // Se faltam apenas 2 a frente, é preciso manter o terceiro numero atrás
      // (let i = totalOfPieces - 4)
      middlePages.push(getGap());
      for (let i = totalOfPieces - 4; i <= totalOfPieces - 2; i++) {
        middlePages.push(colorPaginationOptionAndPutGapIfNeeded(i));
      }
    } else if (totalOfPieces - currentPage <= 5) {
      // Se ainda faltam mais de dois numeros a frente nao e
      // preciso manter 3 numeros atras (let i = currentPage - 2)
      for (let i = currentPage - 2; i <= totalOfPieces - 2; i++) {
        middlePages.push(colorPaginationOptionAndPutGapIfNeeded(i));
      }
      middlePages.unshift(getGap());
    } else if (currentPage <= 2) {
      // Sempre entra aqui ao carregar a pagina, ou quando o usuario volta
      // para a primeira pagina, gera as 4 opcoes a frente
      for (let i = 3; i <= 5; i++) {
        middlePages.push(colorPaginationOptionAndPutGapIfNeeded(i));
      }
      middlePages.push(getGap());
    } else if (currentPage <= 6 && currentPage >= 3) {
      //Mantem as opcoes de paginacao, sempre duas a frente da selecionada
      for (let i = currentPage + 2; i >= 3; i--) {
        middlePages.unshift(colorPaginationOptionAndPutGapIfNeeded(i));
      }
      middlePages.push(colorPaginationOptionAndPutGapIfNeeded('...'));
    } else if (currentPage - 1 >= 6) {
      // Parte do meio, que mantem as retissencias dos dois lados
      middlePages.push(getGap());
      for (let i = 2; i >= -2; i--) {
        middlePages.push(colorPaginationOptionAndPutGapIfNeeded(currentPage - i));
      }
      middlePages.push(getGap());
    }
    return middlePages;
  };

  const createLastOptionPages = () => {
    let lastPages = [];

    if (totalOfPieces >= 4) {
      for (let i = totalOfPieces; i > totalOfPieces - 2; i--) {
        lastPages.unshift(colorPaginationOptionAndPutGapIfNeeded(i));
      }
    } else if (totalOfPieces === 3) {
      lastPages.push(colorPaginationOptionAndPutGapIfNeeded(totalOfPieces));
    }

    return lastPages;
  };

  let paginateOptions = [];
  paginateOptions.push(createInitialOptionPages());
  paginateOptions.push(createMiddleOptionPages());
  paginateOptions.push(createLastOptionPages());

  return (
    <div className="sv-paginate">
      <LinkButton theme="info" disabled={currentPage === 1 || totalOfPieces === 0} onClick={goToPrevious}>
        <i className="fa fa-chevron-left" />
        Anterior
      </LinkButton>
      {paginateOptions}
      <LinkButton theme="info" disabled={totalOfPieces <= currentPage} onClick={goToNext}>
        Próximo
        <i className="fa fa-chevron-right" />
      </LinkButton>
    </div>
  );
};

Paginator.defaultProps = {
  page: 1,
};

Paginator.propTypes = {
  page: PropTypes.number.isRequired,
  limit: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
  onGoToNext: PropTypes.func.isRequired,
  onGoToPrevious: PropTypes.func.isRequired,
  onGoToPage: PropTypes.func.isRequired,
};

export default Paginator;
