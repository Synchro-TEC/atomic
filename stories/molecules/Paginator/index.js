import React from 'react';
// import { text, boolean, object } from "@storybook/addon-knobs";
// import { action } from '@storybook/addon-actions';
// import { optionalSelect } from '../../../utils/optionalSelect';
import { State, Store } from '@sambego/storybook-state';

import readme from './README.md';
import Paginator from '../../../components/molecules/Paginator/Paginator';
import Button from '../../../components/atoms/Button/Button';

const limit = 5;

const store = new Store({
  page: 1,
});

const component = () => {
  function navigate(pager) {
    store.set({ page: pager.currentPage });
  }
  let page = store.get('page');
  return (
    <State store={store}>
      <Paginator
        page={page}
        limit={limit}
        total={21}
        onGoToNext={pager => navigate(pager)}
        onGoToPage={pager => navigate(pager)}
        onGoToPrevious={pager => navigate(pager)}
      />
      <Button theme="info" outline small onClick={() => store.set({ page: 1 })}>
        Reset
      </Button>
    </State>
  );
};

export default [readme, component];
