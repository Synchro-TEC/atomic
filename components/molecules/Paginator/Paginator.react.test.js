import React from 'react';
import Paginator from './Paginator';

describe('Paginator Component', () => {
  test('Render the pagginator', () => {
    const onGoToNextCallBack = jest.fn();
    const onGoToPrevCallBack = jest.fn();
    const onGoToPageCallBack = jest.fn();
    const pagginator = shallow(
      <Paginator
        total={11}
        limit={3}
        page={2}
        onGoToPage={onGoToPageCallBack}
        onGoToNext={onGoToNextCallBack}
        onGoToPrevious={onGoToPrevCallBack}
      />
    );
    expect(pagginator).toMatchSnapshot();
  });

  test('Fire onGoToNext click', () => {
    const onGoToPageCallBack = jest.fn();
    const onGoToNextCallBack = jest.fn();
    const onGoToPrevCallBack = jest.fn();
    const pager = shallow(
      <Paginator
        total={11}
        limit={3}
        onGoToPage={onGoToPageCallBack}
        onGoToNext={onGoToNextCallBack}
        onGoToPrevious={onGoToPrevCallBack}
      />
    );
    pager.find('.next').simulate('click');
    expect(onGoToNextCallBack.mock.calls.length).toEqual(1);
  });

  test('Fire onGoToPrev click', () => {
    const onGoToPageCallBack = jest.fn();
    const onGoToNextCallBack = jest.fn();
    const onGoToPrevCallBack = jest.fn();
    const pager = shallow(
      <Paginator
        total={11}
        limit={3}
        page={2}
        onGoToPage={onGoToPageCallBack}
        onGoToNext={onGoToNextCallBack}
        onGoToPrevious={onGoToPrevCallBack}
      />
    );
    pager.find('.prev').simulate('click');
    expect(onGoToPrevCallBack.mock.calls.length).toEqual(1);
  });

  test('Fire onGoToPage click', () => {
    const onGoToPageCallBack = jest.fn();
    const onGoToNextCallBack = jest.fn();
    const onGoToPrevCallBack = jest.fn();
    const pager = shallow(
      <Paginator
        total={41}
        limit={3}
        page={2}
        onGoToNext={onGoToNextCallBack}
        onGoToPrevious={onGoToPrevCallBack}
        onGoToPage={onGoToPageCallBack}
      />
    );
    pager
      .find('.goto')
      .first()
      .simulate('click', {
        target: { textContent: '2' },
      });
    expect(onGoToPageCallBack.mock.calls.length).toEqual(1);
  });
});
