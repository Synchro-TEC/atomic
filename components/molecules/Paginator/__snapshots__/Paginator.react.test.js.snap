// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator Component Render the pagginator 1`] = `
<div
  className="sv-paginate"
>
  <LinkButton
    className="prev"
    disabled={false}
    onClick={[Function]}
    theme="info"
  >
    <i
      className="fa fa-chevron-left"
    />
    Anterior
  </LinkButton>
  <a
    className="goto"
    key="Jb3S5s6iWvrujFsrrlhV5"
    onClick={[Function]}
  >
     
    1
     
  </a>
  <em
    key="5ryFk~CKlZ5egV8ltfOM2"
  >
     
    2
     
  </em>
  <a
    className="goto"
    key="a06_nSyD0wMwLn0R3S4lL"
    onClick={[Function]}
  >
     
    3
     
  </a>
  <a
    className="goto"
    key="v~d5Glwxm3sdLEHw2ZzIC"
    onClick={[Function]}
  >
     
    4
     
  </a>
  <LinkButton
    className="next"
    disabled={false}
    onClick={[Function]}
    theme="info"
  >
    Pr√≥ximo
    <i
      className="fa fa-chevron-right"
    />
  </LinkButton>
</div>
`;
